documentation/doxygen/man/man3/SettingsWindow.3
.TH "SettingsWindow" 3 "Tue Jan 9 2024" "kOrganizify" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SettingsWindow
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <settingswindow\&.h>\fP
.PP
Inherits QMainWindow\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBchangeColor\fP (QString color)"
.br
.ti -1c
.RI "void \fBsetNotifications\fP (const bool notifications)"
.br
.ti -1c
.RI "bool \fBgetNotifications\fP () const"
.br
.ti -1c
.RI "void \fBupdateNotificationIcon\fP (bool state)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBcolorChanged\fP (const QString &color)"
.br
.ti -1c
.RI "void \fBenabledNotifications\fP (const bool enabled)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSettingsWindow\fP (\fBSettings\fP *settings, QWidget *parent)"
.br
.ti -1c
.RI "\fB~SettingsWindow\fP ()"
.br
.ti -1c
.RI "void \fBsetColor\fP (QString color)"
.br
.ti -1c
.RI "void \fBsetBackgroundPath\fP (QString backgroundPath)"
.br
.ti -1c
.RI "QString \fBgetColor\fP ()"
.br
.ti -1c
.RI "QString \fBgetBackgroundPath\fP ()"
.br
.ti -1c
.RI "QString \fBtextToColor\fP (const QString text)"
.br
.ti -1c
.RI "QString \fBtextToPath\fP (const QString text)"
.br
.ti -1c
.RI "QString \fBcolorToPath\fP (const QString color)"
.br
.ti -1c
.RI "QString \fBcolorToText\fP (QString color)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "Ui::SettingsWindow * \fBui\fP"
.br
.in -1c
.SS "Private Slots"

.in +1c
.ti -1c
.RI "QString \fBgetColorNameFromValue\fP (const QMap< QString, QString > &colorMap, const QString &value)"
.br
.ti -1c
.RI "void \fBsave\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBSettings\fP * \fBm_settings\fP"
.br
.ti -1c
.RI "const QMap< QString, QString > \fBm_themeColors\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SettingsWindow::SettingsWindow (\fBSettings\fP * settings, QWidget * parent)\fC [explicit]\fP"

.SS "SettingsWindow::~SettingsWindow ()"

.SH "Member Function Documentation"
.PP 
.SS "void SettingsWindow::changeColor (QString color)\fC [slot]\fP"

.SS "void SettingsWindow::colorChanged (const QString & color)\fC [signal]\fP"

.SS "QString SettingsWindow::colorToPath (const QString color)"

.SS "QString SettingsWindow::colorToText (QString color)"

.SS "void SettingsWindow::enabledNotifications (const bool enabled)\fC [signal]\fP"

.SS "QString SettingsWindow::getBackgroundPath ()"

.SS "QString SettingsWindow::getColor ()"

.SS "QString SettingsWindow::getColorNameFromValue (const QMap< QString, QString > & colorMap, const QString & value)\fC [private]\fP, \fC [slot]\fP"

.SS "bool SettingsWindow::getNotifications () const\fC [slot]\fP"

.SS "void SettingsWindow::save ()\fC [private]\fP, \fC [slot]\fP"

.SS "void SettingsWindow::setBackgroundPath (QString backgroundPath)"

.SS "void SettingsWindow::setColor (QString color)"

.SS "void SettingsWindow::setNotifications (const bool notifications)\fC [slot]\fP"

.SS "QString SettingsWindow::textToColor (const QString text)"

.SS "QString SettingsWindow::textToPath (const QString text)"

.SS "void SettingsWindow::updateNotificationIcon (bool state)\fC [slot]\fP"

.SH "Member Data Documentation"
.PP 
.SS "\fBSettings\fP* SettingsWindow::m_settings\fC [private]\fP"

.SS "const QMap<QString, QString> SettingsWindow::m_themeColors\fC [private]\fP"
\fBInitial value:\fP
.PP
.nf
{
        {"Blue", "#9EAEF8"},
        {"Green", "#ABD49A"},
        {"Orange", "#F1BC6A"},
        {"Pink", "#FFBCE5"},
        {"Purple", "#D8B7F1"},
        {"Default", "#A5A9A0"}
    }
.fi
.SS "Ui::SettingsWindow* SettingsWindow::ui"


.SH "Author"
.PP 
Generated automatically by Doxygen for kOrganizify from the source code\&.


kOrganizify/src/settingswindow.cpp
#include "settingswindow.h"
#include "ui_settingswindow.h"
#include <QDir>
#include <QMap>

SettingsWindow::SettingsWindow(Settings *settings, QWidget *parent) : QMainWindow(parent), ui(new Ui::SettingsWindow), m_settings(settings) {
    ui->setupUi(this);
    setWindowTitle("Settings");
    setFixedSize(size());


    if (getNotifications()) {
        ui->lblNotificationsOn->setVisible(true);
        ui->lblNotificationsOff->setVisible(false);
    } else {
        ui->lblNotificationsOn->setVisible(false);
        ui->lblNotificationsOff->setVisible(true);
    }

    ui->cbxDropTheme->setCurrentText(colorToText(getColor()));

    connect(ui->cbNotifications, &QCheckBox::stateChanged, this, &SettingsWindow::updateNotificationIcon);
    connect(ui->cbxDropTheme, QOverload<const QString &>::of(&QComboBox::currentTextChanged), [=](const QString &text) {
        QString color = textToColor(text);

        QString path = colorToPath(color);

        setColor(color);
        setBackgroundPath(path);

        QString styleSheet    = QString("background-color: %1; ").arg(color);
        QString btnStyleSheet = QString("QPushButton{" + styleSheet + "border-radius: 10px; color:black;}");
        ui->btnSave->setStyleSheet(btnStyleSheet);
        ui->cbxDropTheme->setStyleSheet(QString("QComboBox{color: black; border-radius:10px; background-color: %1; }").arg(color));
    });

    connect(ui->btnSave, &QPushButton::clicked, this, &SettingsWindow::save);
}

void SettingsWindow::changeColor(QString color) {
    QString styleSheet    = QString("background-color: %1; ").arg(color);
    QString btnStyleSheet = QString("QPushButton{" + styleSheet + "border-radius: 10px; color:black;}");
    ui->btnSave->setStyleSheet(btnStyleSheet);
    ui->cbxDropTheme->setStyleSheet(QString("QComboBox{color: black; border-radius:10px; background-color: %1; }").arg(color));
}

void SettingsWindow::setBackgroundPath(QString backgroundPath) {
    m_settings->setBackgroundPath(backgroundPath);
}

auto SettingsWindow::getBackgroundPath() -> QString {
    return m_settings->backgroundPath();
}

auto SettingsWindow::textToColor(QString text) -> QString {
    return m_themeColors.value(text);
}

auto SettingsWindow::textToPath(QString text) -> QString {
    QString sourceDir   = QCoreApplication::applicationDirPath();
    QString pathBlue    = QDir(sourceDir).filePath(":/resources/images/backgroundBlue.png");
    QString pathGreen   = QDir(sourceDir).filePath(":/resources/images/backgroundGreen.png");
    QString pathOrange  = QDir(sourceDir).filePath(":/resources/images/backgroundOrange.png");
    QString pathPink    = QDir(sourceDir).filePath(":/resources/images/backgroundPink.png");
    QString pathPurple  = QDir(sourceDir).filePath(":/resources/images/backgroundPurple.png");
    QString pathDefault = QDir(sourceDir).filePath(":/resources/images/backgroundDefault.png");

    QMap<QString, QString> themeBackgrounds{{"Blue", pathBlue}, {"Green", pathGreen},   {"Orange", pathOrange},
                                            {"Pink", pathPink}, {"Purple", pathPurple}, {"Default", pathDefault}};

    return themeBackgrounds.value(text, themeBackgrounds["Default"]);
}

auto SettingsWindow::colorToPath(QString color) -> QString {
    return textToPath(colorToText(color));
}

auto SettingsWindow::colorToText(QString color) -> QString {
    return m_themeColors.key(color);
}

void SettingsWindow::save() {
    emit colorChanged(m_settings->getColor());
    emit enabledNotifications(ui->cbNotifications->isChecked());
    close();
}

void SettingsWindow::setColor(const QString color) {
    m_settings->setColor(color);
}

auto SettingsWindow::getColor() -> QString {
    return m_settings->getColor();
}


auto SettingsWindow::getColorNameFromValue(const QMap<QString, QString> &colorMap, const QString &value) -> QString {
    for (auto it = colorMap.begin(); it != colorMap.end(); ++it) {
        if (it.value() == value) {
            return it.key();
        }
    }
    return "";
}

void SettingsWindow::setNotifications(const bool notification) {
    m_settings->setNotifications(notification);
}

auto SettingsWindow::getNotifications() const -> bool {
    return m_settings->getNotifications();
}

void SettingsWindow::updateNotificationIcon(bool state) {
    if (state) {
        ui->lblNotificationsOn->setVisible(true);
        ui->lblNotificationsOff->setVisible(false);
    } else {
        ui->lblNotificationsOn->setVisible(false);
        ui->lblNotificationsOff->setVisible(true);
    }
}

SettingsWindow::~SettingsWindow() {
    delete ui;
}
